fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/PropertiesMap.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * PropertiesMap.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.cascades;
+
+import com.apple.foundationdb.record.query.plan.cascades.expressions.RelationalExpression;
+import com.apple.foundationdb.record.query.plan.cascades.properties.DistinctRecordsProperty;
+import com.apple.foundationdb.record.query.plan.cascades.properties.OrderingProperty;
+import com.apple.foundationdb.record.query.plan.cascades.properties.StoredRecordProperty;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimaps;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.Sets;
+
+import javax.annotation.Nonnull;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Class to manage properties for plans.
+ */
+public class PropertiesMap {
+    private static final Set<PlanProperty<?>> plannerAttributes =
+            ImmutableSet.<PlanProperty<?>>builder()
+                    .add(OrderingProperty.ORDERING)
+                    .add(DistinctRecordsProperty.DISTINCT_RECORDS)
+                    .add(StoredRecordProperty.STORED_RECORD)
+                    .build();
+
+    @Nonnull
+    private final Map<RecordQueryPlan, Map<PlanProperty<?>, ?>> planAttributesMap;
```

Comment:
The `PropertiesMap` is supposed to be the map you describe. What is missing from `PropertiesMap` are better type safe getters and similar things. Will do that.

Reply:
I think it might also benefit from some type-safe `private` methods when dealing with a `PlanProperty<T>` and a `T` at the same time, which is otherwise not there.

Reply:
I added a private `computePropertyValue()` which is now generic by some `P`. None of the other info is leaked out in an unsafe fashion, i.e. all callers either get a typed value via

```
public <P> Map<RecordQueryPlan, P> getPlannerAttributeForAllPlans(@Nonnull final PlanProperty<P> planProperty)
```

or they get data that does not enable them to get to property values (i.e. only `PlanPartition`s and `PlannerProperty` sets (which do not hold their computed values).